src/main/java/org/apache/sysds/runtime/instructions/ooc/MatrixMatrixBinaryOOCInstruction.java

```diff
@@ -0,0 +1,257 @@
+
+public class MatrixMatrixBinaryOOCInstruction extends ComputationOOCInstruction {
+
+
+	protected MatrixMatrixBinaryOOCInstruction(OOCType type, Operator op, CPOperand in1, CPOperand in2, CPOperand out, String opcode, String istr) {
+		super(type, op, in1, in2, out, opcode, istr);
+	}
+
+	public static MatrixMatrixBinaryOOCInstruction parseInstruction(String str) {
+		String[] parts = InstructionUtils.getInstructionPartsWithValueType(str);
+		InstructionUtils.checkNumFields(parts, 4);
+		String opcode = parts[0];
+		CPOperand in1 = new CPOperand(parts[1]); // the 1st large matrix (streamed)
+		CPOperand in2 = new CPOperand(parts[2]); // the 2nd large matrix (streamed)
+		CPOperand out = new CPOperand(parts[3]);
+
+		AggregateOperator agg = new AggregateOperator(0, Plus.getPlusFnObject());
+		AggregateBinaryOperator ba = new AggregateBinaryOperator(Multiply.getMultiplyFnObject(), agg);
+
+		return new MatrixMatrixBinaryOOCInstruction(OOCType.MAPMM, ba, in1, in2, out, opcode, str);
+	}
+
+	/**
+	 *
+	 * @param ec execution context
+	 */
+	@Override
+	public void processInstruction( ExecutionContext ec ) {
+		// 1. Identify the inputs
+		MatrixObject min1 = ec.getMatrixObject(input1);
+		MatrixObject min2 = ec.getMatrixObject(input2);
+
+
+		LocalTaskQueue<IndexedMatrixValue> qIn1 = min1.getStreamHandle();
+		LocalTaskQueue<IndexedMatrixValue> qIn2 = min2.getStreamHandle();
+		LocalTaskQueue<IndexedMatrixValue> qOut = new LocalTaskQueue<>();
+		BinaryOperator plus = InstructionUtils.parseBinaryOperator(Opcodes.PLUS.toString());
+		ec.getMatrixObject(output).setStreamHandle(qOut);
+
+		// Result matrix rows, cols = rows of A, cols of B
+		long resultRowBlocks = min1.getDataCharacteristics().getNumRowBlocks();
+		System.out.println("resultRowBlocks: " + resultRowBlocks);
+		long resultColBlocks = min2.getDataCharacteristics().getNumColBlocks();
+		System.out.println("resultColBlocks: " + resultColBlocks);
+
+		ExecutorService pool = CommonThreadPool.get();
+		try {
+			// Core logic: background thread
+			pool.submit(() -> {
+				IndexedMatrixValue tmpA = null;
+				IndexedMatrixValue tmpB = null;
+				try {
+					// Phase 1: grouping the output blocks by block Index (The Shuffle)
+					Map<MatrixIndexes, List<TaggedMatrixValue>> groupedBlocks = new HashMap<>();
+					HashMap<Long, MatrixBlock> partialResults = new  HashMap<>();
+
+					// Process matrix A: each block A(i,k) contributes to C(i,j) for all j
+					while((tmpA = qIn1.dequeueTask()) != LocalTaskQueue.NO_MORE_TASKS) {
+						long i = tmpA.getIndexes().getRowIndex() - 1;
+						long k = tmpA.getIndexes().getColumnIndex() - 1;
+						System.out.println("i: " + i + ", k: " + k);
+
+						for (int j=0; j<resultColBlocks; j++) {
+							MatrixIndexes index = new MatrixIndexes(i, j); // 1,1= A11,A12,A13,B11,B21,B31
+
+							// Create a copy
+							MatrixBlock sourceBlock = (MatrixBlock) tmpA.getValue();
+							MatrixBlock blockCopy = new MatrixBlock(sourceBlock);
```

Comment:
there is no need for explicit copies - for every block that is streamed in we perform a block matrix multiplication which creates new output blocks and these output blocks are then aggregated if necessary. By default we have copy-on-write-semantics which means operations are never performed in place unless we explicitly say so.